File: 53. SimplifyPath.java
import java.util.ArrayDeque;
import java.util.Deque;

public class SimplifyPath {

    public static String simplifyPath(String path) {
        Deque<String> stack = new ArrayDeque<>();
        for (String part : path.split("/")) {
            if (part.isEmpty() || part.equals(".")) continue;
            if (part.equals("..")) { if (!stack.isEmpty()) stack.pop(); }
            else stack.push(part);
        }
        StringBuilder sb = new StringBuilder();
        for (String p : (Iterable<String>) stack::descendingIterator) sb.append('/').append(p);
        return sb.length() == 0 ? "/" : sb.toString();
    }

    public static void main(String[] args) {
        System.out.println(simplifyPath("/home/"));         // /home
        System.out.println(simplifyPath("/../"));           // /
        System.out.println(simplifyPath("/home//foo/"));    // /home/foo
        System.out.println(simplifyPath("/a/./b/../../c/"));// /c
    }
}

File: 53. SimplifyPath.cpp
#include <iostream>
#include <string>
#include <vector>
#include <sstream>
using namespace std;

string simplifyPath(string path) {
    vector<string> stack;
    stringstream ss(path);
    string part;
    while (getline(ss, part, '/')) {
        if (part.empty() || part == ".") continue;
        if (part == "..") {
            if (!stack.empty()) stack.pop_back();
        } else {
            stack.push_back(part);
        }
    }
    string result;
    for (const string& p : stack) result += '/' + p;
    return result.empty() ? "/" : result;
}

int main() {
    cout << simplifyPath("/home/") << "\n";
    cout << simplifyPath("/../") << "\n";
    cout << simplifyPath("/home//foo/") << "\n";
    cout << simplifyPath("/a/./b/../../c/") << "\n";
    return 0;
}
